vfr: simplify and improve CFR frame dropping algo

jstebbins opened the pull request:
fixes https://github.com/HandBrake/HandBrake/issues/50

jstebbins committed:
vfr: simplify and improve CFR frame dropping algo

The new algo caches extra frames so it can select a "best" frame to drop
from a list. A metric for each buffer is calculated as it is added to
the list. The metric indicates how much the frame differes from the
previous frame. The one with the lowest metric is selected for dropping
when a drop is required.

The old algo tried to latch on to a pattern without keeping any extra
frames. When "in sync" it would drop the buffer that fit the pattern.
But this only worked for a few specific patterns I had tested with and
did not work for all possible patterns (e.g. issue 50 where 1 in 5
needed to be dropped to convert 29.97 to 23.976).

fixes https://github.com/HandBrake/HandBrake/issues/50

bradleysepos commented:
Interesting. Care to elaborate in layman's terms?

jstebbins commented:
The new algo caches extra frames so it can select a "best" frame to drop from a list.  A metric for each buffer is calculated as it is added to the list.  The metric indicates how much the frame differes from the previous frame.  The one with the lowest metric is selected for dropping when a drop is required.

The old algo tried to latch on to a pattern without keeping any extra frames.  When "in sync" it would drop the buffer that fit the pattern.  But this only worked for a few specific patterns I had tested with and did not work for all possible patterns (e.g. issue 50 where 1 in 5 needed to be dropped to convert 29.97 to 23.976).

All that said, I think I see a bug in my assumption of how deep the list needs to be.  There are cases where I may not be keeping a big enough list to reliably select the right frame to drop.  So I need to fix that.

bradleysepos commented:
Cool.

Down the road, I'd like to make temporal NLMeans more intelligent using motion detection to shift the search area. Perhaps that could be made into its own filter like combing detection, so other filters can make use of the data.

sr55 commented:
What do you guys think.  1.1 or 1.0?  
Will it need much testing?

jstebbins commented:
I've tested it on a variety of sources and it is working as intended.  I've also run it through valgrind to verify no leaks.  But at the same time, it's a largish change that could have a corner case I missed. 

I'm fine with pushing this to 1.1 just for the sake of allowing 1.0 to stabilize.

bradleysepos commented:
Seems reasonable to me also to go for 1.1.0.

jstebbins committed:
vfr: add cfr frame drop debugging logs

These logs can be enabled by uncommenting HB_DEBUG_CFR_DROPS at the top
of the file.

If you have any sources that use progressive frame upsampling that you
would like to test with this new frame drop algo, enable these debug
logs to get full details of what frames are dropped and passed.

They show which frames are dropped and the metrics that the decision to
drop is based on.  They also show which frames are passed and the
cadence of passed vs. dropped frames.
